src/workerd/api/r2-bucket.c++

```diff
@@ -541,6 +415,98 @@ R2Bucket::put(jsg::Lock& js, kj::String name, kj::Maybe<R2PutValue> value,
+    createMultipartUploadBuilder.setObject(key);
+
+    KJ_IF_MAYBE(o, options) {
+      KJ_IF_MAYBE(m, o->customMetadata) {
+        auto fields = createMultipartUploadBuilder.initCustomFields(m->fields.size());
+        for (size_t i = 0; i < m->fields.size(); i++) {
+          fields[i].setK(m->fields[i].name);
+          fields[i].setV(m->fields[i].value);
+        }
+      }
+      KJ_IF_MAYBE(m, o->httpMetadata) {
+        auto fields = createMultipartUploadBuilder.initHttpFields();
+        HttpMetadata httpMetadata = [&]() {
+          KJ_SWITCH_ONEOF(*m) {
+            KJ_CASE_ONEOF(m, HttpMetadata) {
+              return kj::mv(m);
+            }
+            KJ_CASE_ONEOF(h, jsg::Ref<Headers>) {
+              return HttpMetadata::fromRequestHeaders(js, *h);
+            }
+          }
+          KJ_UNREACHABLE;
+        }();
+
+        KJ_IF_MAYBE(ct, httpMetadata.contentType) {
+          fields.setContentType(*ct);
+        }
+        KJ_IF_MAYBE(ce, httpMetadata.contentEncoding) {
+          fields.setContentEncoding(*ce);
+        }
+        KJ_IF_MAYBE(cd, httpMetadata.contentDisposition) {
+          fields.setContentDisposition(*cd);
+        }
+        KJ_IF_MAYBE(cl, httpMetadata.contentLanguage) {
+          fields.setContentLanguage(*cl);
+        }
+        KJ_IF_MAYBE(cc, httpMetadata.cacheControl) {
+          fields.setCacheControl(*cc);
+        }
+        KJ_IF_MAYBE(ce, httpMetadata.cacheExpiry) {
+          fields.setCacheExpiry((*ce - kj::UNIX_EPOCH) / kj::MILLISECONDS);
+        }
+      }
+    }
+
+    auto requestJson = json.encode(requestBuilder);
+    auto bucket = adminBucket.map([](auto&& s) { return kj::str(s); });
+
+    auto promise = doR2HTTPPutRequest(js, kj::mv(client), nullptr, nullptr, kj::mv(requestJson),
+        kj::mv(bucket));
+
+    return context.awaitIo(js, kj::mv(promise),
+        [&errorType, key=kj::mv(key), this] (jsg::Lock& js, R2Result r2Result) mutable {
+      r2Result.throwIfError("createMultipartUpload", errorType);
+
+      capnp::MallocMessageBuilder responseMessage;
+      capnp::JsonCodec json;
+      json.handleByAnnotation<R2CreateMultipartUploadResponse>();
+      auto responseBuilder = responseMessage.initRoot<R2CreateMultipartUploadResponse>();
+
+      json.decode(KJ_ASSERT_NONNULL(r2Result.metadataPayload), responseBuilder);
+      kj::String uploadId = kj::str(responseBuilder.getUploadId());
+      return jsg::alloc<R2MultipartUpload>(kj::mv(key), kj::mv(uploadId), JSG_THIS);
+    });
+  });
+}
+
+jsg::Promise<jsg::Ref<R2MultipartUpload>> R2Bucket::resumeMultipartUpload(jsg::Lock& js,
+      kj::String key, kj::String uploadId, const jsg::TypeHandler<jsg::Ref<R2Error>>& errorType) {
+  return js.evalNow([&] {
+    auto& context = IoContext::current();
+    auto promise = kj::Promise<jsg::Ref<R2MultipartUpload>>(jsg::alloc<R2MultipartUpload>(kj::mv(key), kj::mv(uploadId), JSG_THIS));
+    return context.awaitIo(js, kj::mv(promise));
```

Comment:
Returning a promise here instead of a straight R2MultipartUpload to allow us to attach some checks later on. E.g. we might want to send a request to the gateway worker for example to check if this UploadId is valid.

Reply:
That might mislead users into thinking that we actually do those checks though :thinking: I'll need to document explicitly that we don't do those checks.

Reply:
I think you're overthinking things here. A promise is out of place as there would never be any I/O that would be added under the hood.